cgen: too few generated methods in generic arrays/maps inside structs

### Describe the bug

title

### Reproduction Steps

generic, doesnt works:
```v
struct FooParams[K, V] {
	k  K
	v  V
	k2 K
}

struct Foo[K, V] {
	k K
mut:
	m map[K]V
}

fn new_foo[K, V](params FooParams[K, V]) Foo[K, V] {
	return Foo[K, V]{
		k: params.k2
		m: {
			params.k: params.v
		}
	}
}

fn (mut p Foo[K, V]) f() {
	mut y := p.m.move()
	y.clear()
	y.reserve(6)
	y.delete(p.k)
	dump(y.keys())
	dump(y.values())
	dump(y.clone())
	unsafe { y.free() }
}

fn main() {
	mut foo := new_foo(FooParams{ k: 'abc', v: 42, k2: 'def' })
	foo.f()
}
```

non-generic, working version:
```v
fn main() {
    mut x := {
        'abc': 42
    }
    mut y := x.move()
    y.clear()
    y.reserve(6)
    y.delete('k')
    dump(y.keys())
    dump(y.values())
    dump(y.clone())
    unsafe { y.free() }
}
```

### Expected Behavior

successful compile

### Current Behavior

```
==================
C:/Users/mclr/AppData/Local/Temp/v_0/test.01HPM2MGEGJN3PPN9Y59X7T0JG.tmp.c:13249: warning: implicit declaration of function 'Map_string_int_free'
tcc: error: undefined symbol 'Map_string_int_reserve'
tcc: error: undefined symbol 'Map_string_int_free'
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

### Possible Solution

_No response_

### Additional Information/Context

https://github.com/vlang/v/blob/6e22c3592284d5eb4971edd8135981cffd31389c/vlib/v/gen/c/fn.v#L949-L1029

### V version

V 0.4.4 ed5c2f3

### Environment details (OS name and version, etc.)

```
V full version: V 0.4.4 57a33c3.ed5c2f3
OS: windows, Microsoft Windows 11 Pro v22000 64-bit
Processor: 12 cpus, 64bit, little endian, 

getwd: D:\Games\Proekti\JS-TS\disjs
vexe: D:\Games\Proekti\V\v\v.exe
vexe mtime: 2024-02-14 14:54:21

vroot: OK, value: D:\Games\Proekti\V\v
VMODULES: OK, value: C:\Users\mclr\.vmodules
VTMP: OK, value: C:\Users\mclr\AppData\Local\Temp\v_0

Git version: git version 2.37.0.windows.1
Git vroot status: weekly.2024.07-16-ged5c2f31
.git/config present: true

CC version:
thirdparty/tcc status: thirdparty-windows-amd64 a39eb79b
```

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -948,8 +948,16 @@ fn (mut g Gen) gen_arg_from_type(node_type ast.Type, node ast.Expr) {
 
 fn (mut g Gen) gen_map_method_call(node ast.CallExpr, left_type ast.Type, left_sym ast.TypeSymbol) bool {
 	match node.name {
-		'clear' {
-			g.write('map_clear(')
+		'reserve' {
+			g.write('map_reserve(')
+			g.gen_arg_from_type(left_type, node.left)
+			g.write(', ')
+			g.expr(node.args[0].expr)
+			g.write(')')
+			return true
+		}
+		'free', 'clear' {
+			g.write('map_${node.name}(')
 			g.gen_arg_from_type(left_type, node.left)
 			g.write(')')
 			return true

--- a/vlib/v/tests/map_generic_call_test.v
+++ b/vlib/v/tests/map_generic_call_test.v
@@ -1,0 +1,36 @@
+struct FooParams[K, V] {
+	k  K
+	v  V
+	k2 K
+}
+
+struct Foo[K, V] {
+	k K
+mut:
+	m map[K]V
+}
+
+fn new_foo[K, V](params FooParams[K, V]) Foo[K, V] {
+	return Foo[K, V]{
+		k: params.k2
+		m: {
+			params.k: params.v
+		}
+	}
+}
+
+fn (mut p Foo[K, V]) f() {
+	mut y := p.m.move()
+	y.clear()
+	y.reserve(6)
+	y.delete(p.k)
+	assert y.keys().len == 0
+	assert y.values().len == 0
+	assert y.clone().len == 0
+	unsafe { y.free() }
+}
+
+fn test_main() {
+	mut foo := new_foo(FooParams{ k: 'abc', v: 42, k2: 'def' })
+	foo.f()
+}

